src/dom.js

```diff
@@ -0,0 +1,371 @@
+        elm.remove();
+      }
+
+      // This collapses space inside a single textNode..
+      const insertSpaceNext = isText && val.match(/\s$/);
+      if (isText) {
+        elm.nodeValue = elm.nodeValue.trim().replace(/\s+/g, ' ');
+      }
+
+      // This is what applies our cross element collapsing
+      // NOTE: An implementation just collapsing cross element
+      // spaces to the right would be simpler and still correct
+      // in terms of equivalence; it would also be much less
+      // readable when inspected by humans; this is not very pretty:
+      // `Hello <b>World</b>` -> `Hello<b> World</b>`
+      // `<b>Hello</b> <em>World</em>` -> `<b>Hello</b><em> World</em>`
+      if (insertSpace && prev !== undefined) {
+        const prevParents = Array.from(domParentNodes(prev));
+        const elmParents = Array.from(domParentNodes(elm));
+
+        while (prevParents[0].isSameNode(elmParents[0])) {
+          prevParents.shift();
+          elmParents.shift();
+        }
+        const commonParentNode = prevParents[0];
+
+        if (prev.nodeName === '#text' && prevParents.length === 0) {
+          // Like in the `Hello <b>World</b>` case
+          prev.nodeValue += ' ';
+        } else if (isText && elmParents.length === 0) {
+          // Like in the `<b>Hello</b> World` case
+          elm.nodeValue = ' ' + elm.nodeValue;
+        } else {
+          // Like in the `<b>Hello</b> <em>World</em>` case
+          const before = elmParents[0];
+          const commonParent = before.parentNode;
+          const nu = window.document.createTextNode(' ');
+          commonParent.insertBefore(nu, before);
+        }
+      }
+
+      prev = elm;
+      insertSpace = insertSpaceNext;
+    });
+
+    // clear the array; we cannot just reassign the array
+    // because we need to propagate the change to our parent
+    // calls further up in the recursion chain
+    while (inlineTextNodes.length > 0)
+      inlineTextNodes.pop();
+  };
+
+  // normalize() is part of the DOM spec; it merges adjacent textNodes
+  if (root) {
+    node.normalize();
+  }
+
+  if (!inline) {
+    // This tag opens a new block scope!
+    collapseTextAcrossInlineNodes();
+  }
+
+  if (inline && preformatted) {
+    // Entering an inline preformatted node: We push the node
+    // onto the list of textNodes as a marker to notify 
+    // the collapsing function that space around the preformatted
+    // node must not be erased, but instead collapsed to one space
+    // on BOTH sides of the node.
+    // We assign a new list to inlineTextNodes; this allows us to
+    // handle inline, non-preformatted elements inside a preformatted
+    // elements in an isolated environment.
+    inlineTextNodes.push(node);
+    // We assign to inlineTextNodes, but the current list will still
+    // be used up the recursion chain
+    inlineTextNodes = [];
+  }
+
+  each(node.childNodes, (child) => {
+    if (child.nodeName === '#comment') {
+      child.remove();
```

Comment:
does this work in all cases? i.e. if `child.remove()` modifies the original container, the sequence might be off. add a test of 2 consecutive comments to be sure.

Reply:
Iterator stability seems not to be a strong suit of the JS dom :)
The current version iterates over copies of the children lists and has tests for iteratorStability